src/Microsoft.CodeQuality.Analyzers/UnitTests/QualityGuidelines/ReferingToObjectAndReassigningItInTheSameStatementTests.cs

```diff
@@ -0,0 +1,270 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Testing;
+using Microsoft.CodeQuality.Analyzers.QualityGuidelines;
+using Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeQuality.Analyzers.UnitTests.QualityGuidelines
+{
+    public partial class ReferingToObjectAndReassigningItInTheSameStatementTests : DiagnosticAnalyzerTestBase
+    {
+        protected override DiagnosticAnalyzer GetBasicDiagnosticAnalyzer()
+        {
+            return new ReferingToObjectAndReassigningItInTheSameStatement();
+        }
+
+        protected override DiagnosticAnalyzer GetCSharpDiagnosticAnalyzer()
+        {
+            return new ReferingToObjectAndReassigningItInTheSameStatement();
+        }
+
+        private DiagnosticResult GetCSharpResultAt(int line, int column, string symbolName)
+        {
+            return GetCSharpResultAt(line, column, ReferingToObjectAndReassigningItInTheSameStatement.Rule, symbolName);
+        }
+
+        private DiagnosticResult GetBasicResultAt(int line, int column, string symbolName)
+        {
+            return GetBasicResultAt(line, column, ReferingToObjectAndReassigningItInTheSameStatement.Rule, symbolName);
+        }
+
+        [Fact]
+        public void CSharpReassignLocalVariableAndReferToItsField()
+        {
+            VerifyCSharp(@"
+public class C
+{
+    public C Field;
+}
+
+public class Test
+{
+    public void Method()
+    {
+        C a, b;
+        a.Field = a = b;
+    }
+}
+",
+            GetCSharpResultAt(12, 9, "a"));
+        }
+
+        [Fact]
+        public void CSharpReassignLocalVariableAndReferToItsProperty()
+        {
+            VerifyCSharp(@"
+public class C
+{
+    public C Property { get; set; }
+}
+
+public class Test
+{
+    public void Method()
+    {
+        C a, b, c;
+        a.Property = c = a = b;
+    }
+}
+",
+            GetCSharpResultAt(12, 9, "a"));
+        }
+
+        [Fact]
+        public void CSharpReassignLocalVariablesPropertyAndReferToItsProperty()
+        {
+            VerifyCSharp(@"
+public class C
+{
+    public C Property { get; set; }
+}
+
+public class Test
+{
+    public void Method()
+    {
+        C a, b;
+        a.Property.Property = a.Property = b;
```

Comment:
Can you add a test for `b.Property.Property = a.Property = c;` and `b.Property.Property = a.Property = b;`

Reply:
I've added those tests as `CSharpReassignPropertyOfFirstLocalVariableWithSecondAndReferToPropertyOfSecondVariable` and `CSharpReassignPropertyOfFirstLocalVariableWithThirdAndReferToPropertyOfSecondVariable` , and they are failing.
I tried to fix it, but it seems that some sort of reference comparison is needed here, and currently there are only name comparison.
Could you please help me with this?

Reply:
Done. 
You are right, it works! By the way, I noticed, `a.Syntax.ToString() == b.Syntax.ToString()` works for every case without any further conditions. Is there some performance hit for using `Syntax.ToString()`?

Reply:
>  Is there some performance hit for using Syntax.ToString()

Yes, but it should fine given this is not an allocation on a hot path and should be pretty rare.